### PR TITLE
Handle a user cancelling a purchase

### PR DESCRIPTION
**Pull request by ryancoughlin**
### Overview

Starting to work on gracefully handling when a user wants to cancel a purchase. This happens when they hit "Cancel" on the alert before buying.

Rusty with my Obj C, would love some pointers / guidance. The callback does log appropriately:

### Current state of PR
When you tap "Cancel' the error below occurs. The callback is logged correctly:

`2017-07-14 15:22:13.997 [info][tid:com.facebook.React.JavaScript] user_cancelled`

img_0417

### CODE REVIEW

File: InAppUtils/InAppUtils.m [hunk 1]
```diff
@@ -35,6 +35,10 @@ - (void)paymentQueue:(SKPaymentQueue *)queue
                 NSString *key = RCTKeyForInstance(transaction.payment.productIdentifier);
                 RCTResponseSenderBlock callback = _callbacks[key];
                 if (callback) {
+                    if (transaction.error.code == SKErrorPaymentCancelled) {
+                        callback(@[@"user_cancelled"]);
+                        [_callbacks removeObjectForKey:key];
+                    }
                     callback(@[RCTJSErrorFromNSError(transaction.error)]);
```
Comment (ryancoughlin) [inline InAppUtils/InAppUtils.m]: When you don't capture this. The error above is thrown: 
>Cannot connect to the iTunes Store

File: InAppUtils/InAppUtils.m [hunk 2]
```diff
@@ -35,6 +35,10 @@ - (void)paymentQueue:(SKPaymentQueue *)queue
                 NSString *key = RCTKeyForInstance(transaction.payment.productIdentifier);
                 RCTResponseSenderBlock callback = _callbacks[key];
                 if (callback) {
+                    if (transaction.error.code == SKErrorPaymentCancelled) {
```
Comment (cbrevik) [inline InAppUtils/InAppUtils.m]: Would really recommend refactoring out this into its own function like I recommended here: https://github.com/chirag04/react-native-in-app-utils/pull/45#issuecomment-315418161

Comment (cbrevik) [inline InAppUtils/InAppUtils.m]: Also, won't it attempt two callbacks here? The one below the if-block as well?

Comment (cbrevik) [inline InAppUtils/InAppUtils.m]: `RCTJSErrorFromNSError` is still used below, and would presumably still fail for every other error other than cancel?

**Review by cbrevik — COMMENTED** [review]
I think it's a good idea to handle all error cases equally (as I talked about in https://github.com/chirag04/react-native-in-app-utils/pull/45#issuecomment-315418161).

**Comment by ryancoughlin** [comment]
Circling back here, thanks for the early feedback. Its been years since I worked within Obj C so I may not be able to get this finished, but I am willing to push it as far as I can.

File: InAppUtils/InAppUtils.m [hunk 3]
```diff
@@ -6,243 +6,244 @@
...
-        }
-    }
+   for (SKPaymentTransaction *transaction in transactions) {
+       switch (transaction.transactionState) {
+           case SKPaymentTransactionStateFailed: {
+               NSString *key = RCTKeyForInstance(transaction.payment.productIdentifier);
+               RCTResponseSenderBlock callback = _callbacks[key];
+               if (callback) {
+                   NSLog(@"Call back from updatedTransactions: %@", transaction.error);
+                   callback(@[translateErrorCode(transaction.error.code, @"restore_failed")]);
```
Comment (ryancoughlin) [inline InAppUtils/InAppUtils.m]: Similar to this? We would send it the code to switch against and the key we should be returning?

Comment (ryancoughlin) [inline InAppUtils/InAppUtils.m]: I can't seem to get my products to load. This was working yesterday and I haven't changed anything. Not sure if something is up with the App Store right now.

**Comment by ryancoughlin** [comment]
And in case it is related to these changes. Below is what I am getting when my products attempt to load:

```
Error Domain=NSURLErrorDomain Code=-999 "Cannot connect to iTunes Store" UserInfo={NSErrorFailingURLStringKey=https://sandbox.itunes.apple.com/WebObjects/MZInit.woa/wa/initiateSession?ix=6&os=10&locale=en_US, NSLocalizedDescription=Cannot connect to iTunes Store, NSErrorFailingURLKey=https://sandbox.itunes.apple.com/WebObjects/MZInit.woa/wa/initiateSession?ix=6&os=10&locale=en_US}
```

Building to my device and have removed the app and signed out of the App Store several times.

**Comment by ryancoughlin** [comment]
My products are fetching now and my latest logs are:


```
2017-07-18 21:10:33.074 [warn][tid:main][InAppUtils.m:61] No callback registered for transaction with state purchased.
2017-07-18 21:10:33.075 [warn][tid:main][InAppUtils.m:61] No callback registered for transaction with state purchased.
2017-07-18 21:10:33.075 [warn][tid:main][InAppUtils.m:61] No callback registered for transaction with state purchased.
2017-07-18 21:10:33.075 [warn][tid:main][InAppUtils.m:61] No callback registered for transaction with state purchased.
```

Will try and patch up those missing links

**Comment by ryancoughlin** [comment]
@chirag04 Would you be interested in taking a look at this? Not very strong with Obj C and would love to get this resolved to ship this app.

**Pull request closed without merging**